### ISSUE TITLE
Waiting report lists deleted and/or completed tasks

### ISSUE THREAD

**Issue by duckunix** [GothenburgBitFactory/taskwarrior#2626]
#### To report a bug...

* What command(s) did you run?  ```task waiting```
* What did you expect to happen? A list of tasks which are in waiting state but not completed
* What actually happened? A list of tasks which are waiting and some that are completed (tasks are recurring tasks)
* Paste the output of the ```task diag``` command.

```
task 2.6.0
   Platform: Linux

Compiler
    Version: 11.1.0
       Caps: +stdc +stdc_hosted +LP64 +c8 +i32 +l64 +vp64 +time_t64
 Compliance: C++17

Build Features
      CMake: 3.21.3
    libuuid: libuuid + uuid_unparse_lower
  libgnutls: 3.7.2
 Build type: 

Configuration
       File: /home/don/.config/task/taskrc (found), 5055 bytes, mode 100644
       Data: /home/don/.task (found), dir, mode 40755
    Locking: Enabled
         GC: Enabled
    $EDITOR: vim
     Server: xxx.xxxxxxxx.xxx.zzzzz
         CA: /home/don/.task/ca.cert.pem, readable, 3700 bytes
Certificate: /home/don/.task/don_harper.cert.pem, readable, 3710 bytes
        Key: /home/don/.task/don_harper.key.pem, readable, 24742 bytes
      Trust: strict
    Ciphers: NORMAL
      Creds: duckland/Don Harper/************************************

Hooks
     System: Enabled
   Location: /home/don/.task/hooks
     Active: on-add-onenote-manage-notes-annotation.py    (executable)
             on-add.blocks_attr.py                        (executable) (symlink)
             on-exit_dashboard                            (executable) (symlink)
             on-launch.blocks_attr.py                     (executable) (symlink)
             on-modify-onenote-manage-notes-annotation.py (executable)
             on-modify.blocks_attr.py                     (executable)
             on-modify.timewarrior                        (executable)
   Inactive: on-exit                                      (not executable)
             on-modify                                    (not executable)
             task-dashboard                               (executable) (symlink)unrecognized hook name

Tests
   Terminal: 214x30
       Dups: Scanned 7442 tasks for duplicate UUIDs:
             No duplicates found
 Broken ref: Scanned 7442 tasks for broken references:
             No broken references found
```

Output of ```task waiting flame```:
```
ID  Age  P Project      R Wait       Remaining Due        Description
--- ---- - ------------ - ---------- --------- ---------- --------------------------------------------------------
  - 9d   H family.house R 2022-09-24     11mo  2022-10-01 Clean furnace flame sensor
                                                            2020-12-13 https://www.youtube.com/watch?v=tYwR9qL4lnk
                                                            2021-02-09 https://www.youtube.com/watch?v=tYwR9qL4lnk
  - 9d   H family.house R 2022-09-24     11mo  2022-10-01 Clean furnace flame sensor
                                                            2020-12-13 https://www.youtube.com/watch?v=tYwR9qL4lnk
                                                            2021-02-09 https://www.youtube.com/watch?v=tYwR9qL4lnk
304 9d   H family.house R 2022-09-24     11mo  2022-10-01 Clean furnace flame sensor
                                                            2020-12-13 https://www.youtube.com/watch?v=tYwR9qL4lnk
                                                            2021-02-09 https://www.youtube.com/watch?v=tYwR9qL4lnk

3 tasks
```
Details of task 304:
```
Name          Value
------------- -----------------------------------------------------------------
ID            304
Description   Clean furnace flame sensor
                2020-12-13 20:27:48 https://www.youtube.com/watch?v=tYwR9qL4lnk
                2021-02-09 07:08:49 https://www.youtube.com/watch?v=tYwR9qL4lnk
Status        Waiting
Project       family.house
Recurrence    yearly
Parent task   168f5421-7b85-4b46-9eef-a0d32f9f01ac
Mask Index    1.000000
Entered       2021-10-01 00:00:06 (9d)
Waiting until 2022-09-24 00:00:00
Due           2022-10-01 00:00:00
Last modified 2021-02-09 07:08:58 (8mo)
Virtual tags  ANNOTATED CHILD INSTANCE PRIORITY PROJECT UDA UNBLOCKED WAITING
UUID          5270c5ff-bb1d-45ce-bc71-3dc611849bd0
Urgency       15.35
Priority      H

------------------ ------ - ---- - ------
    project             1 *    1 =      1
    waiting             1 *   -3 =     -3
    annotations       0.9 *    1 =    0.9
    due               0.2 *   12 =    2.4
    age             0.025 *    2 =  0.049
    UDA priority.H      1 *    5 =      5
    UDA recur           1 *    4 =      4
    PROJECT family      1 *    5 =      5
                                   ------
                                    15.35
```
Details of one of the closed tasks ```task e4a4c61c-4d2a-40da-a5a3-f494bb010ef5```:
```
Name          Value
------------- ------------------------------------------------------------------------
ID            -
Description   Clean furnace flame sensor (https://www.youtube.com/watch?v=tYwR9qL4lnk)
Status        Completed
Project       family.house
Recurrence    yearly
Parent task   c7403599-0123-4131-b254-a47165cc0438
Mask Index    1.000000
Entered       2019-11-27 20:35:59 (1.9y)
Waiting until 2020-10-05 00:00:00
Due           2020-10-05 22:00:00
End           2020-05-12 23:28:26
Last modified 2020-05-12 23:28:26 (1.4y)
Tags          computer home
Virtual tags  CHILD COMPLETED INSTANCE ORPHAN PRIORITY PROJECT TAGGED UDA UNBLOCKED
UUID          e4a4c61c-4d2a-40da-a5a3-f494bb010ef5
Urgency       29.9
Priority      H

------------------ ------ - ---- - ------
    project             1 *    1 =      1
    tags              0.9 *    1 =    0.9
    due                 1 *   12 =     12
    age                 1 *    2 =      2
    UDA priority.H      1 *    5 =      5
    UDA recur           1 *    4 =      4
    PROJECT family      1 *    5 =      5
                                   ------
                                     29.9
```
The parent task from ```task c7403599-0123-4131-b254-a47165cc0438```
```
Name          Value
------------- ------------------------------------------------------------------------
ID            -
Description   Clean furnace flame sensor (https://www.youtube.com/watch?v=tYwR9qL4lnk)
Status        Deleted
Project       family.house
Recurrence    yearly
Mask          W+
Entered       2019-01-02 21:36:01 (2.8y)
Waiting until 2019-10-05 00:00:00
Due           2019-10-05 22:00:00
End           2019-10-26 00:21:21
Last modified 2020-05-12 23:28:26 (1.4y)
Tags          computer home
Virtual tags  DELETED ORPHAN PARENT PRIORITY PROJECT TAGGED TEMPLATE UDA UNBLOCKED
UUID          c7403599-0123-4131-b254-a47165cc0438
Urgency       29.9
Priority      H

------------------ ------ - ---- - ------
    project             1 *    1 =      1
    tags              0.9 *    1 =    0.9
    due                 1 *   12 =     12
    age                 1 *    2 =      2
    UDA priority.H      1 *    5 =      5
    UDA recur           1 *    4 =      4
    PROJECT family      1 *    5 =      5
                                   ------
                                     29.9

```
Output of ```task waiting -COMPLETED flame```:
```

ID  Age  P Project      R Wait       Remaining Due        Description
--- ---- - ------------ - ---------- --------- ---------- --------------------------------------------------------
  - 9d   H family.house R 2022-09-24     11mo  2022-10-01 Clean furnace flame sensor
                                                            2020-12-13 https://www.youtube.com/watch?v=tYwR9qL4lnk
                                                            2021-02-09 https://www.youtube.com/watch?v=tYwR9qL4lnk
  - 9d   H family.house R 2022-09-24     11mo  2022-10-01 Clean furnace flame sensor
                                                            2020-12-13 https://www.youtube.com/watch?v=tYwR9qL4lnk
                                                            2021-02-09 https://www.youtube.com/watch?v=tYwR9qL4lnk
304 9d   H family.house R 2022-09-24     11mo  2022-10-01 Clean furnace flame sensor
                                                            2020-12-13 https://www.youtube.com/watch?v=tYwR9qL4lnk
                                                            2021-02-09 https://www.youtube.com/watch?v=tYwR9qL4lnk

3 tasks
```

**Comment by duckunix** [comment]
Actually, I mis-typed.  The ```task waiting`` command does suppress completed tasks, but shows *deleted* tasks.

**Comment by duckunix** [comment]
For example:
```
16:18:58 3z [don@loki:~] $ task waiting flame

ID  Age  P Project      R Wait       Remaining Due        Description
  - 9d   H family.house R 2022-09-24     11mo  2022-10-01 Clean furnace flame sensor
                                                            2020-12-13 https://www.youtube.com/watch?v=tYwR9qL4lnk
                                                            2021-02-09 https://www.youtube.com/watch?v=tYwR9qL4lnk
  - 9d   H family.house R 2022-09-24     11mo  2022-10-01 Clean furnace flame sensor
                                                            2020-12-13 https://www.youtube.com/watch?v=tYwR9qL4lnk
                                                            2021-02-09 https://www.youtube.com/watch?v=tYwR9qL4lnk
304 9d   H family.house R 2022-09-24     11mo  2022-10-01 Clean furnace flame sensor
                                                            2020-12-13 https://www.youtube.com/watch?v=tYwR9qL4lnk
                                                            2021-02-09 https://www.youtube.com/watch?v=tYwR9qL4lnk

3 tasks
16:19:04 3z [don@loki:~] $ task waiting status.not:deleted flame

ID  Age  P Project      R Wait       Remaining Due        Description
304 9d   H family.house R 2022-09-24     11mo  2022-10-01 Clean furnace flame sensor
                                                            2020-12-13 https://www.youtube.com/watch?v=tYwR9qL4lnk
                                                            2021-02-09 https://www.youtube.com/watch?v=tYwR9qL4lnk

1 task
```

Deleted task 
```
16:20:32 3z [don@loki:~] $ task 817b87de-3cbe-4b6f-8612-b4254e37b490

Name          Value
ID            -
Description   Clean furnace flame sensor
                2020-12-13 20:27:48 https://www.youtube.com/watch?v=tYwR9qL4lnk
                2021-02-09 07:08:49 https://www.youtube.com/watch?v=tYwR9qL4lnk
Status        Deleted
Project       family.house
Recurrence    yearly
Parent task   168f5421-7b85-4b46-9eef-a0d32f9f01ac
Mask Index    1.000000
Entered       2021-10-01 00:00:01 (9d)
Waiting until 2022-09-24 00:00:00
Due           2022-10-01 00:00:00
End           2021-10-03 00:29:12
Last modified 2021-10-03 00:29:12 (7d)
Virtual tags  ANNOTATED CHILD DELETED INSTANCE PRIORITY PROJECT UDA UNBLOCKED WAITING
UUID          817b87de-3cbe-4b6f-8612-b4254e37b490
Urgency       15.35
Priority      H

    project             1 *    1 =      1
    waiting             1 *   -3 =     -3
    annotations       0.9 *    1 =    0.9
    due               0.2 *   12 =    2.4
    age             0.025 *    2 =  0.049
    UDA priority.H      1 *    5 =      5
    UDA recur           1 *    4 =      4
    PROJECT family      1 *    5 =      5
                                   ------
                                    15.35
```

**Comment by duckunix** [comment]
So, I did have it write the first time, but also with the deleted tasks.  *sigh*

```
16:26:09 3z [don@loki:~] $ task waiting Kay

ID  Age  Project Recur Due   Description                Urg
298 3w   family  P365D 11mo  Order bday present for Kay 9.54
  - 3w   family  P365D 11mo  Order bday present for Kay 9.54

2 tasks
16:26:21 3z [don@loki:~] $ task waiting status.not:completed Kay

ID  Age  Project Recur Due   Description                Urg
298 3w   family  P365D 11mo  Order bday present for Kay 9.54

1 task
16:26:31 3z [don@loki:~] $ task 1118e6b8-3498-4593-ba23-6d2468061ebe

Name          Value
ID            -
Description   Order bday present for Kay
Status        Completed
Project       family
Recurrence    yearly
Parent task   f61f6b62-dd00-436a-a771-db02c07dadad
Mask Index    2.000000
Entered       2021-09-15 00:00:10 (3w)
Waiting until 2022-09-01 00:00:00
Due           2022-09-15 00:00:00
End           2021-09-23 11:05:35
Last modified 2021-09-23 11:05:35 (2w)
Virtual tags  CHILD COMPLETED INSTANCE PROJECT UNBLOCKED WAITING
UUID          1118e6b8-3498-4593-ba23-6d2468061ebe
Urgency       9.537

    project             1 *    1 =      1
    waiting             1 *   -3 =     -3
    due               0.2 *   12 =    2.4
    age             0.068 *    2 =  0.137
    UDA recur           1 *    4 =      4
    PROJECT family      1 *    5 =      5
                                   ------
                                    9.537
```

**Comment by tbabej** [comment]
Good find! I'm assuming you have not customized the `waiting` report and are using the default value of the filter for the waiting report, like this?

```
$ task show report.waiting.filter

Config Variable       Value   
report.waiting.filter +WAITING
```

If that is so, you can recover the same behaviour by reconfiguring the report to only include pending tasks using `task config report.waiting.filter "+PENDING +WAITING"`

This is definitely a bug, though I'm not sure how to fix it. Simplest fix would be to change the default definition of the `waiting` report's filter to the definition suggested above, but perhaps the task should not be considered waiting if it's completed or deleted.

**Comment by duckunix** [comment]
Correct, I do not have a custom waiting report when I see this behavior.

I can get around it by adding:
```
report.waiting.filter     =status.not:deleted status.not:completed +WAITING
```

Trying the filter you suggested returns no tasks for me.

**Comment by tbabej** [comment]
Ah, right. My suggestion was stupid, because in 2.6.0 for compatibility reasons we kept the `status:waiting` around if task has `wait` attribute in the future. Since status:waiting by definition excludes status:pending, +PENDING did not match anything in combination with +WAITING. The status:waiting will be removed in the next major release (3.0), and only THEN the suggestion I had would work. I was living too much in the future :slightly_smiling_face: 

TL;DR:
* right now, defining `report.waiting.filter` as `status:waiting` should also fix this problem (but this solution will break in 3.0)
* your workaround works now, and will work in future releases, including 3.0 and later (though the `status.not` parts might become unnecessary once we fix this bug, which should happen in 2.6.1)

**Comment by duckunix** [comment]
Cool...thanks for the update.  I look forward to our new task future. :smile:

**Comment by tbabej** [comment]
This should be now fixed (and tested) on the develop branch, hopefully 2.6.1 release will soon follow :slightly_smiling_face:

**Comment by ChargingBulle** [comment]
This is also very important for `tasksh` users.

The default review criterium includes everything with `+WAITING`:

```
report._reviewed.description=Tasksh review report.  Adjust the filter to your needs.
report._reviewed.columns=uuid
report._reviewed.sort=reviewed+,modified+
report._reviewed.filter=( reviewed.none: or reviewed.before:now-6days ) and ( +PENDING or +WAITING )
```

**Comment by tbabej** [comment]
@TuringZauber great point! Let me just chime in here that in the meantime 2.6.1 got released, so upgrading should make this problem go away :+1:
